lib/mysql/mysql.go

```diff
@@ -0,0 +1,757 @@
+ * Usage:
+ *	var sql *mysql.Connection := mysql.NewConnection(&mysql.Config{
+ *		Host: targetHost,
+ *		Port: targetPort,
+ *	})
+ *	err := sql.Connect()
+ *	defer sql.Disconnect()
+ * The Connection exports the connection details via the ConnectionLog.
+ */
+package mysql
+
+import (
+	"bufio"
+
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/zmap/zcrypto/tls"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// Start state
+	STATE_NOT_CONNECTED = "NOT_CONNECTED"
+
+	// After the TCP connection is completed
+	STATE_CONNECTED = "CONNECTED"
+
+	// After reading a HandshakePacket with SSL capabilities, before sending the SSLRequest packet
+	STATE_SSL_REQUEST = "SSL_REQUEST"
+
+	// After sending an SSLRequest packet, before peforming an SSL handshake
+	STATE_SSL_HANDSHAKE = "SSL_HANDSHAKE"
+
+	// After connection has been negotiated (from either CONNECTED or SSL_HANDSHAKE)
+	STATE_FINISHED = "STATE_FINISHED"
+)
+
+type ConnectionState string
+
+const (
+	PACKET_TYPE_OK          = "OK"
+	PACKET_TYPE_ERROR       = "ERROR"
+	PACKET_TYPE_HANDSHAKE   = "HANDSHAKE"
+	PACKET_TYPE_EOF         = "EOF"
+	PACKET_TYPE_SSL_REQUEST = "SSL_REQUEST"
+)
+
+// Capability flags: See https://dev.mysql.com/doc/dev/mysql-server/8.0.1/group__group__cs__capabilities__flags.html
+const (
+	CLIENT_LONG_PASSWORD uint32 = (1 << iota)
+	CLIENT_FOUND_ROWS
+	CLIENT_LONG_FLAG
+	CLIENT_CONNECT_WITH_DB
+	CLIENT_NO_SCHEMA
+	CLIENT_COMPRESS
+	CLIENT_ODBC
+	CLIENT_LOCAL_FILES
+	CLIENT_IGNORE_SPACE
+	CLIENT_PROTOCOL_41
+	CLIENT_INTERACTIVE
+	// CLIENT_SSL = (1<<11)
+	CLIENT_SSL
+	CLIENT_IGNORE_SIGPIPE
+	CLIENT_TRANSACTIONS
+	CLIENT_RESERVED
+	// CLIENT_SECURE_CONNECTION = (1 << 15)
+	CLIENT_SECURE_CONNECTION
+	CLIENT_MULTI_STATEMENTS
+	CLIENT_MULTI_RESULTS
+	CLIENT_PS_MULTI_RESULTS
+	// CLIENT_PLUGIN_AUTH = ( 1 << 19 )
+	CLIENT_PLUGIN_AUTH
+	CLIENT_CONNECT_ATTRS
+	CLIENT_PLUGIN_AUTH_LEN_ENC_CLIENT_DATA
+	CLIENT_CAN_HANDLE_EXPIRED_PASSWORDS
+	CLIENT_SESSION_TRACK
+	CLIENT_DEPRECATED_EOF
+)
+
+// Config defaults
+const (
+	DEFAULT_TIMEOUT_SECS        = 3
+	DEFAULT_PORT                = 3306
+	DEFAULT_CLIENT_CAPABILITIES = CLIENT_SSL
+	DEFAULT_RESERVED_DATA_HEX   = "0000000000000000000000000000000000000000000000"
+)
+
+type Config struct {
+	// @TODO: Does it make sense to make Host/Port connection fields, so that Config can be shared across connections?
+	Host string
+	Port uint16
+
+	TLSConfig          *tls.Config
+	Timeout            time.Duration
+	ClientCapabilities uint32
+	MaxPacketSize      uint32
+	CharSet            byte
+	ReservedData       []byte
+}
+
+// flagsToList() converts an integer flags variable to a list of consts corresponding to each bit.
+// The [i]th entry of consts corresponds to the [i]th bit in flags.
+// Example: flagsToList(0x11, { "a", "b", "c", "d", "e" }) returns { "a", "e" }.
+func flagsToList(flags uint64, consts []string) (ret []string) {
```

Comment:
This is slightly different than what we discussed -- I had originally said it would be a "set" (map of const -> true, like `{"FLAG_1": true, "FLAG_X": true }`), but this is implemented as an array (`["FLAG_1", "FLAG_X"]`).

For using the flags in code, the set is definitely more convenient, but at a glance, I find the list easier to read -- and I'm guessing it would be easier for queries as well.

Reply:
I'd definitely do a map. It's very difficult to query lists in a database.